Add command rewrite design doc (#7)
wpilibsuite/design-docs: CommandRewriteDesignDoc.md

# WPILib Command Framework Rewrite
Design document

The WPILib Command-based framework has been a staple of FRC Java and C++ robot code development for years, offering teams a relatively clean and sensible API for factoring complex robot code into simple, encapsulated chunks that interact in clear, comprehensible ways.  In addition to helping teams avoid the often substantial overhead of rolling their own state machine logic, it also serves as an important pedagogical tool for teaching teams about encapsulation, control flow, and design patterns.

However, the libraries used in the framework are now more than a decade old.  Many of the particular choices in their implementation are obsolete, and others are messy, difficult-to-maintain, and/or confusingly-conceived.  Given the confusing nature of much of the implementation, bringing the API up to modern standards is a difficult task, and the libraries have ended up in a "stasis" of sorts.

To document will describe the goals, design choices, and implementation of a ground-up rewrite of the Command-based libraries aimed at alleviating these issues.

> ## Why rewrite the framework? (Problem Summary)

Reviewer: Can you please rewrite this section to be less condemning of the old framework? Teams have successfully used it for 10+ years and I think writing about it in this way is not good for it or WPILib. See some of my inline comments.

> The problems with the current command framework fall generally into three major categories:

Reviewer: Off by one error?
```suggestion
The areas for improvement with the current command framework fall generally into four major categories:
```

> 1.  Poor readability/maintainability.

Reviewer: ```suggestion
1.  Maintainability
```

> 2.  Poor encapsulation/separation of responsibilities.

Reviewer: ```suggestion
2.  Encapsulation & separation of responsibilities
```

> 3.  Unnecessarily restrictive API design.

Reviewer: ```suggestion
3.  Restrictive API design
```

> 4.  Clutter.

Reviewer: ```suggestion
4.  API Clutter
```

> These three problems are interrelated, but I will summarize each in turn.

Reviewer: ```suggestion
These four areas are interrelated, but I will summarize each in turn.
```

### Problem 4: The API Is Really, Really Cluttered

Because of deeper issues with separation-of-concerns between the Command-based libraries and the `Sendable` implementation, many `Command`s have long lists of overloaded constructors and other cruft that make the code difficult to navigate and provide little useful functionality to the users.  The `PIDCommand` class is a particularly egregious example.

Overall, the problems with the library run deep enough that an attempt to modify the existing structure would likely be futile.  Large, breaking changes are needed to address the concerns raised above.

## Overall Design of the Rewrite

> The rewritten Command-based framework seeks to rectify the above problems.  A broad overview of the changes introduced in the rewrite is as follows:

Reviewer: ```suggestion
The rewritten Command-based framework seeks to improve on the areas listed above.  A broad overview of the changes introduced in the rewrite is as follows:
```

##### whileActiveOnce vs whileActiveContinuous

The `whileActive()` method previously re-scheduled the command continuously while the trigger was active.  This has now been named `whileActiveContinuous()`, and a `whileActiveOnce()` method has been added that cancels the command when the trigger becomes inactive, but does not re-start it if it finishes while the trigger is still active.  The corresponding `Button` method names are `whileHeld()` and `whenHeld()`.

##### Runnable bindings

The `whenActive`, `whileActiveContinuous`, and `whenInactive` bindings (and their corresponding `Button` equivalents) have been given the ability to take a `Runnable`, which is wrapped in an `InstantCommand` automatically.  This allows the use of the edge-finding logic for tasks that the user does not wish to write a command for.

##### Trigger composition

```java
Trigger and(Trigger trigger);
Trigger or(Trigger trigger);
Trigger negate();
```

These allow `Trigger`s to be composed prior to button binding, a la:

```java
button1.and(button2).whenActive(...);
```

> 

Reviewer: Extra EOL.
```suggestion
```
